### ISSUE TITLE
GitHub plugin does not send OAuth token when fetching diff data for PR

### ISSUE THREAD

**Issue by nyankoframe** [gitextensions/gitextensions#9117]
## Current behaviour

Certain GitHub organizations that I contribute to use SAML/single-sign on and also have custom visibility settings set following the [GitHub innersource](https://resources.github.com/whitepapers/introduction-to-innersource) model. These organizations require an authorized PAT be sent in the HTTP header in order to access any repositories. 

The GitHub plugin sends the PAT in the HTTP header for viewing the list of PRs and comments for PR, but does not send the PAT when querying for diff data in the PR, resulting in a 404 return code from GitHub.

## Expected behaviour
GitHub plugin sends the PAT when querying for diff data in PR, regardless of whether the repository/organization requires it.

## Steps to reproduce

1. Ensure that you are a member of at least one organization that uses custom visibility settings and that you have a PAT entered in GitHub plugin settings
2. Choose "View Pull Requests" from the GitHub menu
3. Choose a repository from the custom-visibility organization and wait for the list of PRs to load. You may also see a "Not Found" error message box during this step which is also caused by the missing PAT in the HTTP headers.

## Screenshots

I'm not able to provide any screenshots currently for this issue.

## Did this work in previous version of GitExtensions

Unknown.

## Environment

- Git Extensions 3.4.3.9999
- Build d4b0f48bbf3e39a71f5d7ff5231be5678d4aa0f1
- Git 2.31.0.windows.1
- Microsoft Windows NT 10.0.19042.0
- .NET Framework 4.8.4341.0
- DPI 120dpi (125% scaling)

## Diagnostics

Used [Fiddler](https://fiddler2.com) to decrypt HTTPS traffic and view request and response headers for the following operations:
* Fetching the list of PRs from a custom-visibility organization
* Fetching diff data for a PR
* Fetching comments for a PR

**Comment by RussKie** [comment]
Pull requests are welcome

**Comment by nyankoframe** [comment]
> Pull requests are welcome

Do you / other contributors have a preference for how this could be fixed?

* Credentials should be added via HttpWebRequest.Headers.Add
* Define a new method in the Git.hub external dependency

**Comment by RussKie** [comment]
@pmiossec you probably know more than anyone else about this area

**Comment by pmiossec** [comment]
Yes, I have seen this issue but I don't know enough the code and I have to have a look first.

**Comment by pmiossec** [comment]
@nyankoframe 
In `Git.Hub.PullRequest`, add:

```cs
        public string GetDiffData(string diffUrl)
        {
            var request = new RestRequest(diffUrl);
            return _client.Get<string>(request).Content;
        }
```

and replace `GitHubPullRequest.GetDiffDataAsync()` by:

```cs
        public string GetDiffData()
        {
            if (_diffData is null)
            {
                _diffData = _pullRequest.GetDiffData(_pullRequest.DiffUrl);
            }

            return _diffData;
        }
```

Adapt call to this method and that should do the job...

**Comment by RussKie** [comment]
Can `_client.Get<string>(request)` result in `null`?

**Comment by pmiossec** [comment]
> Can `_client.Get<string>(request)` result in `null`?

I'm not sure but I don't think.
It was a quick hack to give a way to follow but yes, we need to handle the errors well.
That's the main reason why I didn't do the PR myself with the fix (also because changes have to been made in the submodule)
There is some more work needed, indeed...
